### PR TITLE
[SPARK-41036][CONNECT][PYTHON] `columns` API should use `schema` API to avoid data fetching

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

Current `columns` is implemented based on `limit` which runs a job to fetch data and get schema from the data collection. However a more efficient way is to call `schema` API which only need to analyze the plan without collect data. This approach should be more efficient in most of the cases.

### Why are the changes needed?

Efficiency

### Does this PR introduce _any_ user-facing change?

NO

### How was this patch tested?

UT

### CODE REVIEW

**Comment by amaliujia** [comment]
R: @zhengruifeng @HyukjinKwon

File: python/pyspark/sql/connect/dataframe.py [hunk 1]
```diff
@@ -139,11 +139,9 @@ def columns(self) -> List[str]:
         if self._plan is None:
             return []
         if "columns" not in self._cache and self._plan is not None:
-            pdd = self.limit(0).toPandas()
```
Comment (zhengruifeng) [inline python/pyspark/sql/connect/dataframe.py:142]: maybe we don't need to cache `columns`, just return `self.schema.names`?

Comment (amaliujia) [reply python/pyspark/sql/connect/dataframe.py:142]: I would prefer to not depend on the underly API when doing caching...

E.g. what if someday the cache on the schema is gone but this API is not aware of it, etc.

Comment (HyukjinKwon) [reply python/pyspark/sql/connect/dataframe.py:142]: Hmmm .. why do we need `_cache`? I think we can just remove.

File: python/pyspark/sql/connect/dataframe.py [hunk 2]
```diff
@@ -139,11 +139,9 @@ def columns(self) -> List[str]:
...
-            if pdd is None:
-                raise Exception("Empty result")
+            names = self.schema().names
```
Comment (HyukjinKwon) [inline python/pyspark/sql/connect/dataframe.py:142]: This LGTM

File: python/pyspark/sql/connect/dataframe.py (hunk 1 above)
Comment (amaliujia) [reply python/pyspark/sql/connect/dataframe.py:142]: hmm if users call this API multiple times, we only need one gRPC. This should be useful right?

Comment (amaliujia) [reply python/pyspark/sql/connect/dataframe.py:142]: Ok if you think this is a bit over-engineering I can remove.

Comment (HyukjinKwon) [reply python/pyspark/sql/connect/dataframe.py:142]: For that case, we should probably have a proper cache layer instead of doing this alone in `names`.

Comment (HyukjinKwon) [reply python/pyspark/sql/connect/dataframe.py:142]: we can do that for all cases for metadata-ish cases. e.g., schema or even collected results

Comment (amaliujia) [reply python/pyspark/sql/connect/dataframe.py:144]: I removed this caching stuff in this PR. After we pretty much support enough API, we can go back to build a cache layer for all metadata like API to save RPC calls.

**Review by zhengruifeng — APPROVED** [review]

**Comment by zhengruifeng** [comment]
merged into master

**Pull request closed without merging**